{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3e8fd583",
   "metadata": {},
   "source": [
    "<div style=\"text-align: justify; padding:5px; background-color:rgb(252, 253, 255); border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;\">\n",
    "    <font color='red'>Mini Jupyter tutorial<br><br>To run each cell, click the cell and press <kbd>Run</kbd> from the menu bar. This will run any Python code or display any text within the selected cell before highlighting the next cell down. There are two types of cell: A <i>text cell</i> of type <kbd>Markdown</kbd> or <kbd>Heading</kbd> and a <i>code cell</i> of type <kbd>Code</kbd> identifiable with the <span style=\"font-family: courier; color:black; background-color:white;\">In[ ]:</span> to the left of the cell</i>. The type of cell is also identifiable from the dropdown menu in the above menu bar to the right of <kbd>Run</kbd>. Any visual results produced by the code (text/figures) are displayed directly below that cell. Press <kbd>Run</kbd> again until you reach the end of the notebook or alternatively click <kbd>Kernel</kbd><font color='black'>→</font><kbd>Restart and Run All</kbd>. Should the Jupyter notebook crash for any reason, restart the Jupyter Kernel by clicking <kbd>Kernel</kbd><font color='black'>→</font><kbd>Restart</kbd>, and start again from the top.\n",
    "        \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d445be5",
   "metadata": {},
   "source": [
    "# Comparison on variants between individuals to generate a couples report\n",
    "\n",
    "<p style=\\\"text-align: justify\\\">\n",
    "<br>\n",
    "    This workflow parses and compares two annotated variant CSV files and generates a couples report\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b6f80791",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from natsort import index_natsorted\n",
    "from number_parser import parse_ordinal\n",
    "import os\n",
    "import glob"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a0e35da",
   "metadata": {},
   "source": [
    "<div style=\"background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;\">\n",
    "\n",
    "<h3 style=\"text-align: justify\">Given a list of couples CSV file in folder <kbd>input_couples_csv</kbd>, indivduals are compared and a report CSV is generated for each couple</h3>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "64299494",
   "metadata": {},
   "outputs": [],
   "source": [
    "couples = pd.read_csv('input_couples_csv/couples.csv')\n",
    "\n",
    "pathogenic_panel = ['Pathogenic', 'Likely pathogenic', 'Suspected pathogenic', 'Assess further']\n",
    "core_gene_panel = ['CFTR', 'SMN1', 'HBA1', 'HBA2', 'HBA1/HBA2', 'HBB']\n",
    "additional_finding_gene_panel = ['ATM', 'LDLR', 'FH', 'TTN']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8056f367",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort the final report by patient ID and gene before writing out\n",
    "def custom_sort(col: pd.Series) -> pd.Series:\n",
    "    if col.name == \"Patient ID\":\n",
    "        return col.apply(parse_ordinal)\n",
    "    return col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "04c0c8fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = os.getcwd()\n",
    "\n",
    "for index, couple in couples.iterrows():\n",
    "    individual_1_file = couple.iloc[0]\n",
    "    individual_2_file = couple.iloc[1]\n",
    "\n",
    "    individual_1_file_path = glob.glob(os.path.join(path, \"output_annotated_csv/*\"+individual_1_file+\"*.csv\"))[0]\n",
    "    individual_2_file_path = glob.glob(os.path.join(path, \"output_annotated_csv/*\"+individual_2_file+\"*.csv\"))[0]\n",
    "\n",
    "    individual_1 = pd.read_csv(individual_1_file_path)\n",
    "    individual_2 = pd.read_csv(individual_2_file_path)\n",
    "    \n",
    "    ind1 = individual_1[['Gene', 'Transcript', 'Start', 'Variant', 'ClassificationOverall']]\n",
    "    ind2 = individual_2[['Gene', 'Transcript', 'Start', 'Variant', 'ClassificationOverall']]\n",
    "    \n",
    "    # Identify common variants that are pathogenic between couples\n",
    "    couple_variants = pd.merge(ind1, ind2, left_on=['Gene', 'Transcript', 'Start', 'Variant', 'ClassificationOverall'], right_on=['Gene', 'Transcript', 'Start', 'Variant', 'ClassificationOverall']).drop_duplicates()\n",
    "\n",
    "    couple_condition = [(couple_variants.ClassificationOverall.isin(pathogenic_panel))]\n",
    "\n",
    "    couple_choices = ['Potential Couple at Risk']\n",
    "\n",
    "    couple_variants['Reason'] = np.select(couple_condition, couple_choices)\n",
    "    couple_variants = couple_variants[couple_variants.Reason.isin(couple_choices)]\n",
    "\n",
    "    couple_gene_risk = list(couple_variants.Gene)\n",
    "\n",
    "    couple_transcript_risk = list(couple_variants.Transcript)\n",
    "\n",
    "    couple_variant_risk = list(couple_variants.Variant)\n",
    "\n",
    "    couple_classify_risk = list(couple_variants.ClassificationOverall)\n",
    "    \n",
    "    couple = [individual_1, individual_2]\n",
    "\n",
    "    report = pd.DataFrame()\n",
    "\n",
    "    #Filter and assess variants from each couple\n",
    "    for individual in couple:\n",
    "    \n",
    "        genes = np.unique(individual['Gene'])\n",
    "        gene_variant_classify = individual[['Gene', 'Variant', 'ClassificationOverall']].drop_duplicates().reset_index(drop=True)\n",
    "    \n",
    "        #Filter gene variants to only pathogenic and assess further cases\n",
    "        gene_variant_classify = gene_variant_classify[gene_variant_classify.ClassificationOverall.isin(pathogenic_panel)]\n",
    "    \n",
    "        gene_variant = gene_variant_classify[['Gene', 'Variant']].drop_duplicates().reset_index(drop=True)\n",
    "    \n",
    "        genes = np.unique(gene_variant['Gene'])\n",
    "        \n",
    "        #Count the number of pathogenic and assess further variants per gene \n",
    "        df_gene_variant_grouping = pd.DataFrame()\n",
    "        for gene in genes:\n",
    "            gene_variant_grouping = gene_variant[gene_variant.Gene == gene].copy(deep=True)\n",
    "            gene_variant_grouping['Gene_Variant_Count'] = len(gene_variant_grouping)       \n",
    "\n",
    "            if df_gene_variant_grouping.empty:\n",
    "                df_gene_variant_grouping = gene_variant_grouping\n",
    "            else:\n",
    "                df_tmp = gene_variant_grouping\n",
    "                df_gene_variant_grouping = pd.concat([df_gene_variant_grouping, df_tmp], axis=0).reset_index(drop=True)\n",
    "    \n",
    "        #Merge the counts back into the full dataset, then set NaN gene_variant_count values to 0\n",
    "        merged_individual = individual.merge(df_gene_variant_grouping, left_on=['Gene', 'Variant'], right_on=['Gene', 'Variant'], how='outer').drop_duplicates().reset_index(drop=True)\n",
    "        \n",
    "        merged_individual['Gene_Variant_Count'] = merged_individual['Gene_Variant_Count'].fillna(0)\n",
    "        \n",
    "        #Filter variants based on series of conditions\n",
    "        conditons_dict = dict({})\n",
    "        \n",
    "        conditons_dict['Potential Affected (HOM)'] = [(merged_individual.Zygosity == 'HOMOZYGOTE')\n",
    "                                        & (merged_individual.ClassificationOverall.isin(pathogenic_panel))]\n",
    "    \n",
    "        conditons_dict['Potential Affected (COMPHET)'] = [((merged_individual.Gene_Variant_Count >= 2)\n",
    "                                        & (merged_individual.ClassificationOverall.isin(pathogenic_panel)))]    \n",
    "    \n",
    "        conditons_dict['Potential X-linked'] = [(merged_individual.Chromosome == 'X')\n",
    "                                        & (merged_individual.ClassificationOverall.isin(pathogenic_panel))]\n",
    "    \n",
    "        conditons_dict['Core Gene'] = [(merged_individual.Gene.isin(core_gene_panel))\n",
    "                                    & (merged_individual.ClassificationOverall.isin(pathogenic_panel))]\n",
    "    \n",
    "        conditons_dict['Potential Additional Finding'] = [(merged_individual.Gene.isin(additional_finding_gene_panel))\n",
    "                                    & (merged_individual.ClassificationOverall.isin(pathogenic_panel))]\n",
    "    \n",
    "        conditons_dict['Potential Couple at Risk'] = [(merged_individual.Gene.isin(couple_gene_risk))\n",
    "                                                   & (merged_individual.Transcript.isin(couple_transcript_risk))\n",
    "                                                   & (merged_individual.Variant.isin(couple_variant_risk))\n",
    "                                                   & (merged_individual.ClassificationOverall.isin(couple_classify_risk))]\n",
    "                                                  \n",
    "        # Build the variant assessment report\n",
    "        for key, value in conditons_dict.items():\n",
    "            \n",
    "            choices = [key]\n",
    "            \n",
    "            sample_id_choices = [list(individual['Sample#'])] * len(choices)\n",
    "            genomic_coordinates = [list(individual['Genomic Coordinates'])] * len(choices)\n",
    "            ref = [list(individual['Ref'])] * len(choices)\n",
    "            alt = [list(individual['Alt'])] * len(choices)\n",
    "            gene = [list(individual['Gene'])] * len(choices)\n",
    "            variant = [list(individual['Variant'])] * len(choices)\n",
    "            variant_protein = [list(individual['Protein Change'])] * len(choices)\n",
    "            transcript = [list(individual['Transcript'])] * len(choices)        \n",
    "            allele_frequency = [[x[0].split('=')[1] for x in list(individual['Variant QC'].str.split(';'))]] * len(choices)\n",
    "            zygosity = [list(individual['Zygosity'])] * len(choices)\n",
    "            classification_overall = [list(individual['ClassificationOverall'])] * len(choices)\n",
    "            \n",
    "            if report.empty:\n",
    "                report['Patient ID'] = np.select(value, sample_id_choices)\n",
    "                report['Genomic Coordinates'] = np.select(value, genomic_coordinates)\n",
    "                report['Ref'] = np.select(value, ref)\n",
    "                report['Alt'] = np.select(value, alt)\n",
    "                report['Gene'] = np.select(value, gene)\n",
    "                report['Variant Change'] = np.select(value, variant)\n",
    "                report['Transcript'] = np.select(value, transcript)\n",
    "                report['Variant Protein'] = np.select(value, variant_protein)\n",
    "                report['Zygosity'] = np.select(value, zygosity)\n",
    "                report['Allele Frequency'] = np.select(value, allele_frequency)\n",
    "                report['ClassificationOverall'] = np.select(value, classification_overall)\n",
    "                report['Reason'] = np.select(value, choices)\n",
    "                report = report.loc[~(report=='0').all(axis=1)].reset_index(drop=True)\n",
    "            else:\n",
    "                df_tmp = pd.DataFrame()\n",
    "                df_tmp['Patient ID'] = np.select(value, sample_id_choices)\n",
    "                df_tmp['Genomic Coordinates'] = np.select(value, genomic_coordinates)\n",
    "                df_tmp['Ref'] = np.select(value, ref)\n",
    "                df_tmp['Gene'] = np.select(value, ref)\n",
    "                df_tmp['Alt'] = np.select(value, alt)\n",
    "                df_tmp['Gene'] = np.select(value, gene)\n",
    "                df_tmp['Variant Change'] = np.select(value, variant)\n",
    "                df_tmp['Transcript'] = np.select(value, transcript)\n",
    "                df_tmp['Variant Protein'] = np.select(value, variant_protein)\n",
    "                df_tmp['Zygosity'] = np.select(value, zygosity)\n",
    "                df_tmp['Allele Frequency'] = np.select(value, allele_frequency)\n",
    "                df_tmp['ClassificationOverall'] = np.select(value, classification_overall)\n",
    "                df_tmp['Reason'] = np.select(value, choices)  \n",
    "                report = pd.concat([report, df_tmp], axis=0)\n",
    "                report = report.loc[~(report=='0').all(axis=1)].reset_index(drop=True)\n",
    "        \n",
    "    report['Analyst comment'] = [''] * len(report)\n",
    "\n",
    "    report = report.sort_values(by=['Patient ID', 'Gene'], key=custom_sort)\n",
    "    \n",
    "    # Check if there is an empty report with no variants, then add comment\n",
    "    if len(report) == 0:\n",
    "        report['Analyst comment'] = ['No Variants for Reporting']\n",
    "    \n",
    "    output_file = \"output_couples_report/\"+ individual_1_file +\"_\"+ individual_2_file +\"_couples_report.csv\"\n",
    "    report.to_csv(output_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22f45ef3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
